src/Agent.Sdk/Util/ParentProcessUtil/InteropParentProcessFinder.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace Agent.Sdk.Util.ParentProcessUtil
+{
+    internal class Interop
+    {
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct PROCESS_BASIC_INFORMATION
+        {
+            internal IntPtr ExitStatus;
+            internal IntPtr PebBaseAddress;
+            internal IntPtr AffinityMask;
+            internal IntPtr BasePriority;
+            internal IntPtr UniqueProcessId;
+            internal IntPtr InheritedFromUniqueProcessId;
+        }
+
+        [DllImport("ntdll.dll", SetLastError = true)]
+        internal static extern int NtQueryInformationProcess(IntPtr processHandle, int processInformationClass, out PROCESS_BASIC_INFORMATION processInformation, int processInformationLength, out int returnLength);
+    }
+
+    internal static class InteropParentProcessFinder
+    {
+        private const string ParentProcessState = nameof(ParentProcessState);
+
+        internal static int? GetParentProcessId(IntPtr handle)
+        {
+            Interop.PROCESS_BASIC_INFORMATION pbi;
+            int returnLength;
+            int status = Interop.NtQueryInformationProcess(handle, 0, out pbi, Marshal.SizeOf<Interop.PROCESS_BASIC_INFORMATION>(), out returnLength);
+
+            if (status != 0)
+            {
+                return null;
+            }
+
+            int parentProcessId = pbi.InheritedFromUniqueProcessId.ToInt32();
+
+            return parentProcessId;
+        }
+
+        internal static (Process process, Dictionary<string, string> telemetry) GetParentProcess(Process process)
+        {
+            var telemetry = new Dictionary<string, string>();
+
+            try
+            {
+                int? parentProcessId = GetParentProcessId(process.Handle);
+
+                if (parentProcessId == null)
+                {
+                    return (null, telemetry);
+                }
+
+                Process parentProcess = Process.GetProcessById(parentProcessId.Value);
+
+                if (parentProcess == null
+                    || parentProcess.StartTime > process.StartTime
```

Comment:
not obvious why we do this

Reply:
This was added to check if the process can really be a parent one, because Windows may reuse process ids